### PR TITLE
ASoC: SOF: ipc4-pcm: fix dma_id for capture

### PR DESCRIPTION
**Pull request by plbossart**
The existing code uses (stream_tag - 1) for the host and link dma id.

This is correct for playback, but for capture this results in an invalid dma_type being used. The firmware assumes that the dma_id for capture is always larger than DAI_NUM_HDA_OUT

FIXME: check how we can store num_capture and remove the hard-coded '9' value.

### CODE REVIEW

**Comment by plbossart** [comment]
pointer to the firmware code here: https://github.com/thesofproject/sof/blob/8d2fb32fa50156d6c6361d6ecd16905d6e51cf73/src/audio/chain_dma.c#L410

**Comment by plbossart** [comment]
the topology PR https://github.com/thesofproject/sof/pull/8076 is required to test this on HDaudio platforms

File: sound/soc/sof/ipc4-pcm.c [hunk 1]
```diff
@@ -294,6 +296,15 @@ static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
 			msg.extension |= pipeline->msg.extension;
 	}
 
+	if (direction == SNDRV_PCM_STREAM_CAPTURE) {
+		/*
+		 * for capture we need to add num_playback to the value (stream_tag - 1)
+		 * stored in msg.primary
+		 */
+		msg.primary +=  SOF_IPC4_GLB_CHAIN_DMA_HOST_ID(ipc4_data->num_playback_streams);
+		msg.primary +=  SOF_IPC4_GLB_CHAIN_DMA_LINK_ID(ipc4_data->num_playback_streams);
```
Comment (bardliao) [inline sound/soc/sof/ipc4-pcm.c:310]: Do we need to check if dma host id and dma link id exceeds the max id after adding num_playback?

Comment (ujfalusi) [reply sound/soc/sof/ipc4-pcm.c:310]: Why don't do these where we set the DMA ids in the first place?
sof_ipc4_prepare_copier_module() - for HOST_ID
sof_ipc4_dai_config() - for LINK_ID

Comment (ujfalusi) [reply sound/soc/sof/ipc4-pcm.c:310]: Right, sof_ipc4_dai_config() does not  know the direction :(

File: sound/soc/sof/ipc4-pcm.c [hunk 2]
```diff
@@ -294,6 +296,15 @@ static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
 			msg.extension |= pipeline->msg.extension;
 	}
 
+	if (direction == SNDRV_PCM_STREAM_CAPTURE) {
+		/*
+		 * for capture we need to add num_playback to the value (stream_tag - 1)
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c]: I would reword it to something like this:
```c
		/*
		 * For ChainDMA the DMA ids are unique with the following mapping:
		 * playback:  0 - (num_playback_streams - 1)
		 * capture:   num_playback_streams - (num_playback_streams +
		 *				      num_capture_streams - 1)
		 *
		 * Add the num_playback_streams offset to the DMA ids stored in
		 * msg.primary in case capture
		 */
```

File: sound/soc/sof/ipc4-pcm.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-pcm.c:310]: > Do we need to check if dma host id and dma link id exceeds the max id after adding num_playback?

@bardliao, I would expect that we already done checks when the stream_tag was picked?

**Review by ujfalusi — COMMENTED** [review]
@plbossart, my only nitpick is the comment, I would be more precise with it

File: sound/soc/sof/ipc4-pcm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-pcm.c:310]: we could be paranoid indeed, but there is already limits on stream_tag already. I am with @ujfalusi it's probably fine.

File: sound/soc/sof/ipc4-pcm.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-pcm.c]: thanks for the suggestion, will add.

Comment (plbossart) [reply sound/soc/sof/ipc4-pcm.c]: done

File: sound/soc/sof/ipc4-pcm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-pcm.c:310]: no change.

File: sound/soc/sof/ipc4-priv.h [hunk 3]
```diff
@@ -66,6 +66,8 @@ struct sof_ipc4_fw_library {
  * @nhlt: NHLT table either from the BIOS or the topology manifest
  * @mtrace_type: mtrace type supported on the booted platform
  * @mtrace_log_bytes: log bytes as reported by the firmware via fw_config reply
+ * @num_playback_streams: max number of playback DMAs, needed for CHAIN_DMA offset
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-priv.h:69]: nitpicks:
1. commit message:
```
- The CHAIN_DMA IPC needs the number of playback streams as an offset for the dma_id.
+ The CHAIN_DMA IPC needs the number of playback streams as a start offset for the dma_id of a capture stream.
```
2. Is this documented in any way for the IPC4 protocol? We just need to be clear on this for other vendors who will try to implement ChainDMA.

**Review by ujfalusi — COMMENTED** [review]
@plbossart, looks nice, it is unfortunate that we have this awkwardly defined offset issue for capture dma_id.
No much we can do about it...

One nitpick and one note for the first patch for your consideration.

File: sound/soc/sof/ipc4-priv.h (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-priv.h:69]: updated commit message as suggested.

I also added a note 
````
The CHAIN_DMA IPC is currently only functional on Intel HDaudio DMAs,
    and gated by the snd_sof_is_chain_dma_supported() helper.
````

**Review by ujfalusi — APPROVED** [review]

**Review by bardliao — APPROVED** [review]
